Modules/_functoolsmodule.c

```diff
@@ -280,30 +280,33 @@ partial_new(PyTypeObject *type, PyObject *args, PyObject *kw)
         assert(PyTuple_Check(pto->args));
     }
 
+    /* process keywords */
     if (pto_kw == NULL || PyDict_GET_SIZE(pto_kw) == 0) {
-        if (kw == NULL) {
-            pto->kw = PyDict_New();
-        }
-        else if (Py_REFCNT(kw) == 1) {
-            pto->kw = Py_NewRef(kw);
-        }
-        else {
-            pto->kw = PyDict_Copy(kw);
-        }
+        pto->kw = PyDict_New();
     }
     else {
         pto->kw = PyDict_Copy(pto_kw);
-        if (kw != NULL && pto->kw != NULL) {
-            if (PyDict_Merge(pto->kw, kw, 1) != 0) {
-                Py_DECREF(pto);
-                return NULL;
-            }
-        }
     }
     if (pto->kw == NULL) {
         Py_DECREF(pto);
         return NULL;
     }
+    if (kw != NULL) {
+        PyObject *key, *val;
+        Py_ssize_t pos = 0;
+        while (PyDict_Next(kw, &pos, &key, &val)) {
+            if (val == phold) {
+                Py_DECREF(pto);
+                PyErr_SetString(PyExc_TypeError,
+                                "keyword Placeholders are not allowed");
+                return NULL;
+            }
+            if (PyDict_SetItem(pto->kw, key, val)) {
+                Py_DECREF(pto);
+                return NULL;
+            }
+        }
```

Comment:
The logic here is either (1) _error out if placeholder found_ or (2) _copy the entire `kwargs` into a new dict_.

Why not first validate and then use `PyDict_Copy()`? This will avoid inserting items one by one (which is arguably slower and harder to understand).

Reply:
It was using `PyDict_Copy` / `PyDict_Merge` before.

But this way it proved to be faster. So I thought if I can do both at the same time and it is faster, then why not.

Unless there are some good reasons not to do it this way?

Reply:
Could it be because the test samples you used did not include a large dict?

If it proved to be faster for even a large enough kw object, then this looks good to me.

Reply:
It seems that all of these are calling `setitem_take2_lock_held` under the hood.

`PyDict_Merge/Update` have slight overhead, but nothing major. Mostly, because it not only handles `dict` as argument 2, but any object with `keys` and `__getitem__`.

So it is slightly faster to do it in 1 loop explicitly.

Reply:
I cannot find "mark as resolved" button for my comments. Please consider them as resolved.